### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [3]:
#new df, calculations, print
purchase_df = purchase_data[["SN", "Age", "Gender", "Item Name", "Price"]]
players = purchase_df["SN"].unique()
number_players = len(players)
print(number_players)

576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
#calculations
items = purchase_df["Item Name"].unique()
number_items = len(items)
average_price = round(purchase_df["Price"].mean(), 2)
number_purchases = purchase_df["Price"].count()
total_revenue = purchase_df["Price"].sum()

#summary table
PA_summary_table = pd.DataFrame({"Number of Unique Items": [number_items], "Average Price": [average_price], "Total Revenue": [total_revenue]})                            
PA_summary_table["Total Revenue"] = PA_summary_table["Total Revenue"].map("${:.2f}".format)
print(PA_summary_table)

   Number of Unique Items  Average Price Total Revenue
0                     179           3.05      $2379.77


## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
#drop duplicates
demo_df = purchase_df.drop_duplicates(['SN','Age','Gender'])

#calculations using value_counts
gender_count = demo_df["Gender"].value_counts()
gender_percent = gender_count / number_players

#summary table
Demo_gender_summary_df = pd.DataFrame({"Total Count": gender_count, "Percentage of Players": gender_percent})

#format
Demo_gender_summary_df["Percentage of Players"] = Demo_gender_summary_df["Percentage of Players"].map("{:.2%}".format)

#display results
Demo_gender_summary_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [6]:
#groupby and calculations
grouped_gender_df = purchase_data.groupby(['Gender'])
gender_purchases = grouped_gender_df.count()["Price"]
gender_mean_price = grouped_gender_df.mean()["Price"]
gender_total_revenue = grouped_gender_df.sum()["Price"]
gender_avg_total_revenue = gender_total_revenue / Demo_gender_summary_df["Total Count"]

#new data frame
PA_gender_analysis = pd.DataFrame({"Purchase Count": gender_purchases,
                                   "Average Purchase Price": gender_mean_price,
                                  "Total Purchase Value": gender_total_revenue,
                                  "Avg Total Purchase per Person": gender_avg_total_revenue})
#format
PA_gender_analysis["Average Purchase Price"] = PA_gender_analysis["Average Purchase Price"].map("${:.2f}".format)
PA_gender_analysis["Total Purchase Value"] = PA_gender_analysis["Total Purchase Value"].map("${:.2f}".format)
PA_gender_analysis["Avg Total Purchase per Person"] = PA_gender_analysis["Avg Total Purchase per Person"].map("${:.2f}".format)

#display results
PA_gender_analysis

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [7]:
#bin and bin labels
age_bins = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
bins = [0, 10, 15, 20, 25, 30, 35, 40, 99999]

#pd cut to add new column
pd.cut(demo_df["Age"], bins, right=True, labels=age_bins, include_lowest=False)
demo_df["Age Range"] = pd.cut(demo_df["Age"], bins, right=False, labels=age_bins, include_lowest=True)

#calculations
age_count = demo_df["Age Range"].value_counts()
age_percent = age_count / number_players

#Sort, format and print summary
Age_demo_summary = pd.DataFrame({"Total Count": age_count, "Percentage of Players": age_percent})
Age_demo_summary = Age_demo_summary.sort_index()
Age_demo_summary["Percentage of Players"] = Age_demo_summary["Percentage of Players"].map("{:.2%}".format)
Age_demo_summary

<ipython-input-7-75b8a84efd60>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo_df["Age Range"] = pd.cut(demo_df["Age"], bins, right=False, labels=age_bins, include_lowest=True)


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
# bins and bin labels
age_bins = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
bins = [0, 10, 15, 20, 25, 30, 35, 40, 99999]
#pd.cut to add new column
purchase_data["Age Range"] = pd.cut(purchase_data["Age"], bins, right=False, labels=age_bins, include_lowest=True)
purchase_data.head()

#group by on age range
group_age= purchase_data.groupby("Age Range")

#try value_counts in place of groupby
age_purchases = group_age["Price"].count()
age_avg_price = group_age["Price"].mean()
age_total_purchases = group_age["Price"].sum()
avg_purchase_per_person = age_total_purchases / Age_demo_summary["Total Count"]

#create my summary dataframe
#Print summary
PA_Age_summary = pd.DataFrame({"Purchase Count": age_purchases,
                               "Average Purchase Price": age_avg_price,
                               "Total Purchase Value": age_total_purchases,
                               "Avg Total Purchase Per Person": avg_purchase_per_person})
#format
PA_Age_summary["Average Purchase Price"] = PA_Age_summary["Average Purchase Price"].map("${:.2f}".format)
PA_Age_summary["Total Purchase Value"] = PA_Age_summary["Total Purchase Value"].map("${:.2f}".format)
PA_Age_summary["Avg Total Purchase Per Person"] = PA_Age_summary["Avg Total Purchase Per Person"].map("${:.2f}".format)

#display results
PA_Age_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Range                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05             $1114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase Per Person  
Age Range                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
#calculations using groupby
top_spenders_df = purchase_data.groupby(['SN'])
top_purchases = top_spenders_df.count()["Price"]
top_avg_price = top_spenders_df.mean()["Price"]
top_total_purchases = top_spenders_df.sum()["Price"]
top_avg_total_revenue = top_total_purchases / Demo_gender_summary_df["Total Count"]

#Summary table
TopSpenders_summary_table_df = pd.DataFrame({"Purchase Count": top_purchases,
                                             "Average Purchase Price": top_avg_price,
                                             "Total Purchase Value": top_total_purchases})
#sort from highest to lowest
TopSpenders_summary_table_df = TopSpenders_summary_table_df.sort_values("Total Purchase Value", ascending=False)

#format
TopSpenders_summary_table_df["Total Purchase Value"] = TopSpenders_summary_table_df["Total Purchase Value"].map("${:.2f}".format)
TopSpenders_summary_table_df["Average Purchase Price"] = TopSpenders_summary_table_df["Average Purchase Price"].map("${:.2f}".format)

#display results
TopSpenders_summary_table_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [18]:
#groupby [Item ID, Item Name]
popular_items_df = purchase_data.groupby(["Item ID", "Item Name"])

#Calculations
pop_purchase_count = popular_items_df.count()["Purchase ID"]
pop_avg_price = popular_items_df.mean()["Price"]
pop_total = popular_items_df.sum()["Price"]

#Create summary table
Popular_items_summary_df = pd.DataFrame({"Purchase Count": pop_purchase_count,
                                             "Item Price": pop_avg_price,
                                             "Total Purchase Value": pop_total})
#sort highest to lowest
Popular_items_summary_df = Popular_items_summary_df.sort_values("Purchase Count", ascending=False)

#display
Popular_items_summary_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                      Item Price  \
Item ID Item Name                                                  
92      Final Critic                                    4.614615   
178     Oathbreaker, Last Hope of the Breaking Storm    4.230000   
145     Fiery Glass Crusader                            4.580000   
132     Persuasion                                      3.221111   
108     Extraction, Quickblade Of Trembling Hands       3.530000   

                                                      Total Purchase Value  
Item ID Item Name                                                           
92      Final Critic                                                 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
132     Persuasion                                                   28.99  
108     Extraction, Quickblade Of Trembling Hands                    31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [19]:
#sort total purhcase value - highest to lowest
Most_profitable_summary_df = Popular_items_summary_df.sort_values("Total Purchase Value", ascending=False)

#display
Most_profitable_summary_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                      Item Price  \
Item ID Item Name                                                  
92      Final Critic                                    4.614615   
178     Oathbreaker, Last Hope of the Breaking Storm    4.230000   
82      Nirvana                                         4.900000   
145     Fiery Glass Crusader                            4.580000   
103     Singed Scalpel                                  4.350000   

                                                      Total Purchase Value  
Item ID Item Name                                                           
92      Final Critic                                                 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
103     Singed Scalpel                                               34.80

Trend Analysis:

Gender: A large majority of players (84.03%) are male; and account for the majority of sales, by count and value.  However, the average per person purchase totals was not significant based on gender.

Age: The majority of players (44.79%) are between the ages of 20-24, with another 18.58% between the ages of 15-19. These two age groups also account for the most purchases, however there is age-based trend for the average per person spending based on age.

Spending: Per person purchase count and total purchase values do not fluctuate greatly from highest to lowest. In other words, per player spending is low.

Most Popular and Most Profitable: There appears to be a positive correlation between these two categories, sharing the same top two and three of five items.  The top two items have significantly higher sales counts than other items.